src/BlazorWebView/src/SharedSource/AutoCloseOnReadCompleteStream.cs

```diff
@@ -0,0 +1,48 @@
+﻿#if WEBVIEW2_WINFORMS || WEBVIEW2_WPF
+
+using System.IO;
+
+namespace Microsoft.AspNetCore.Components.WebView.WebView2
+{
+	internal class AutoCloseOnReadCompleteStream : Stream
+	{
+		private readonly Stream _baseStream;
+
+		public AutoCloseOnReadCompleteStream(Stream baseStream)
+		{
+			_baseStream = baseStream;
+		}
+
+		public override bool CanRead => _baseStream.CanRead;
+
+		public override bool CanSeek => _baseStream.CanSeek;
+
+		public override bool CanWrite => _baseStream.CanWrite;
+
+		public override long Length => _baseStream.Length;
+
+		public override long Position { get => _baseStream.Position; set => _baseStream.Position = value; }
+
+		public override void Flush() => _baseStream?.Flush();
+
+		public override int Read(byte[] buffer, int offset, int count)
+		{
+			var bytesRead = _baseStream.Read(buffer, offset, count);
+
+			if (bytesRead == 0)
+			{
+				_baseStream.Close();
+			}
```

Comment:
I do not think this check is correct. You might receive a `0` count or an empty buffer and those will result in `_baseStream.Read` returning 0.

Would it work if instead we compare the current position against the length of the stream? We should be able to leverage here the fact that we are dealing with files on disk.

Reply:
Hmm that's what I initially had, but it necessitated a bit more complexity to avoid accessing the properties of a closed out stream. I did this based on:

> The total number of bytes read into the buffer. This can be less than the number of bytes requested if that many bytes are not currently available, or zero (0) if the end of the stream has been reached.

https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.read?view=net-6.0#system-io-stream-read(system-byte()-system-int32-system-int32)

> Implementations return the number of bytes read. The implementation will block until at least one byte of data can be read, in the event that no data is available. [Read](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.read?view=net-6.0) returns 0 only when there is no more data in the stream and no more is expected (such as a closed socket or end of file). An implementation is free to return fewer bytes than requested even if the end of the stream has not been reached.

Based on this, I believe this should be fine as-is?

Reply:
I guess the empty buffer is still potentially problematic here, but that almost seems like an issue in the callee. Maybe we just throw an exception in this case?

Reply:
Could you not keep track of whether you closed the underlying stream already? Would it also not throw before you get to this check in the call to Read if the internal stream was already closed?

Reply:
I am somewhat fine with the additional info provided; those are the types of things that are great comments to add in the code to give context on why that might be Ok to do.

Are there other APIs on the stream that a caller can call instead of just Read that might bypass the check?

Reply:
> Are there other APIs on the stream that a caller can call instead of just Read that might bypass the check?

As far as I can tell this should be sufficient as it'll serve as the marker for the end-of-stream is reached on read. I've done some basic testing locally and things are working as expected.

cc/ @yildirimcagri this is a workaround we're implementing for https://github.com/MicrosoftEdge/WebView2Feedback/issues/2513. Essentially wrapping the stream and closing it out when the read is completed. Wanted to get your thoughts on this approach and see if you had any concerns?